<a href="https://colab.research.google.com/github/mori-mkm/UFSCar/blob/main/PAE/Tukey%E2%80%99s%20Test/exemplo_competicao_vinho.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Importando as bibliotecas

Primeiro, vamos carregar os pacotes e funções necessários em Python:

In [1]:
import pandas as pd
import numpy as np
from scipy.stats import f_oneway
from statsmodels.stats.multicomp import pairwise_tukeyhsd

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


# Inserindo os dados e ajustando o modelo

O código a seguir mostra como criar um conjunto com quatro grupos (A, B, C e D) e ajustar um modelo ANOVA unilateral aos dados para determinar se os valores médios de cada grupo são iguais:

In [4]:
#enter data for three groups
a = [20,	15,	18,	26,	22,	19]
b = [24,	18,	19,	26,	24,	21]
c = [28,	23,	24,	30,	28,	27]
d = [28,	24,	23,	30,	26,	25]
#perform one-way ANOVA
f_oneway(a, b, c, d)


F_onewayResult(statistic=6.478873239436619, pvalue=0.0030548164401503638)

Podemos ver que o valor-p da tabela ANOVA é 0.003055.

Como isso é menor que 0,05, temos evidências suficientes para dizer que os valores médios em cada grupo não são iguais.

Assim, podemos prosseguir para realizar o Teste de Tukey para determinar exatamente quais médias de grupo são diferentes.

# Teste de Tukey

Para realizar o teste de Tukey em Python, podemos usar a função **pairwise_tukeyhsd()** da biblioteca statsmodels:

In [5]:
#create DataFrame to hold data
df = pd.DataFrame({'score': [20,	15,	18,	26,	22,	19,
                             24,	18,	19,	26,	24,	21,
                             28,	23,	24,	30,	28,	27,
                             28,	24,	23,	30,	26,	25],
                   'group': np.repeat(['a', 'b', 'c', "d"], repeats=6)}) 

# perform Tukey's test
tukey = pairwise_tukeyhsd(endog=df['score'],
                          groups=df['group'],
                          alpha=0.05)

#display results
print(tukey)

Multiple Comparison of Means - Tukey HSD, FWER=0.05 
group1 group2 meandiff p-adj   lower   upper  reject
----------------------------------------------------
     a      b      2.0 0.6623 -2.9722  6.9722  False
     a      c   6.6667 0.0063  1.6944 11.6389   True
     a      d      6.0 0.0146  1.0278 10.9722   True
     b      c   4.6667 0.0707 -0.3056  9.6389  False
     b      d      4.0 0.1435 -0.9722  8.9722  False
     c      d  -0.6667    0.9 -5.6389  4.3056  False
----------------------------------------------------


Veja como interpretar o resultado:

* Valor-P para a diferença de médias entre a e b: 0.662

* Valor-P para a diferença de médias entre a e c: 0.0063  

* Valor-P para a diferença de médias entre a e d: 0.0146   

* Valor-P para a diferença nas médias entre b e c: 0.0707

* Valor-P para a diferença nas médias entre b e d: 0.1435

* Valor-P para a diferença nas médias entre c e d:  0.9

Assim, poderíamos concluir que existe uma diferença estatisticamente significativa entre as médias dos grupos a e c; a e d; b e c. Mas não uma diferença estatisticamente significativa entre as médias dos grupos a e b; b e d; c e d.